### PR TITLE
Try load custom exceptions via class loader on client

### PR DESCRIPTION
**Pull request by sancar**
We are throwing UndefinedErrorCodeException if an exception is
not on the protocol list.

This causes poor experience as the behaviour is different between
the client and the member.

see  https://github.com/hazelcast/hazelcast/issues/9753

This pr does not introduce an ExceptionFactory API as discussed
on the issue.
The value of an ExceptionFactory API is debetable and left out
for now. If the client has the expcetion class on the classpath,
the client will create it. If it is not available on the classpath,
it is not clear what can a user do with ExceptionFactory API.
In that case, we are throwing UndefinedErrorCodeException as before.
The main problem seems to be the case where the client have the
exact same class on the classpath, so this fix should cover
most of the use cases.

Also added  assert to check if exception is defined in the protocol
When classLoading is introduced it is possible for us to forget
to put the exception in the protocol, because it will not longer
throw UndefinedErrorCodeException but it will be loaded
automatically.
Assertion is added to check and warn us.

fixes https://github.com/hazelcast/hazelcast/issues/9753

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2474/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2474/Hazelcast-pr-builder-2474.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ClientRegressionWithMockNetworkTest.testGithubIssue3557:673 ? TestTimedOut tes...
[ERROR]   ClientAtomicLongTest.alterAndGetAsync_whenException:291 ? TestTimedOut test ti...
[ERROR]   ClientAtomicLongTest.alterAndGet_whenException:275 ? TestTimedOut test timed o...
[ERROR]   ClientAtomicLongTest.alterAsync_whenException:228 ? TestTimedOut test timed ou...
[ERROR]   ClientAtomicLongTest.alter_whenException:212 ? TestTimedOut test timed out aft...
[ERROR]   ClientAtomicLongTest.applyAsync_whenException:187 ? TestTimedOut test timed ou...
[ERROR]   ClientAtomicLongTest.apply_whenException:172 ? TestTimedOut test timed out aft...
[ERROR]   ClientAtomicLongTest.getAndAlterAsync_whenException:351 ? TestTimedOut test ti...
[ERROR]   ClientAtomicLongTest.getAndAlter_whenException:335 ? TestTimedOut test timed o...
[INFO] 
[ERROR] Tests run: 4822, Failures: 0, Errors: 9, Skipped: 292
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Review by puzpuzpuz — APPROVED** [review]
LGTM with the same nits as in #17556

File: hazelcast-client/src/test/java/com/hazelcast/client/impl/protocol/ClientExceptionFactoryTest.java [hunk 1]
```diff
@@ -187,7 +176,6 @@ private boolean stackTraceArrayEquals(StackTraceElement[] stackTrace1, StackTrac
                 new Object[]{new EntryProcessorException(randomString())},
                 new Object[]{new ArrayIndexOutOfBoundsException(randomString())},
                 new Object[]{new ArrayStoreException(randomString())},
-                new Object[]{new AuthenticationException(randomString())},
```
Comment (ihsandemir) [inline hazelcast-client/src/test/java/com/hazelcast/client/impl/protocol/ClientExceptionFactoryTest.java:190]: why did you remove?

Comment (sancar) [reply hazelcast-client/src/test/java/com/hazelcast/client/impl/protocol/ClientExceptionFactoryTest.java:190]: On 3.12, because we had two different modules, we are throwing `StubAuthenticationException` on the member and it is recreated as `AuthenticationException` on the client-side. 
So this test normally should not be passing. When I refactored the test to a correct state, this issue became apparent, and tests start to fail for `AuthenticationException`. 
I did not want to modify the test again for the `AuthenticationException` only. 
Now thinking about it, I can put a special test for `AuthenticationException` on `ClientProtocolErrorCodesTest`.

**Review by ihsandemir — APPROVED** [review]
